benchmark/metrics_parser.py

```diff
@@ -0,0 +1,276 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+
+import argparse
+from collections import defaultdict
+import csv
+import itertools
+import json
+
+
+WALL_CLOCK_KEY = u'-- WALL --'
+
+
+HEADERS = [
+    u'Flocker Version',
+    u'Nodes',
+    u'Containers',
+    u'Scenario',
+    u'Control Service CPU',
+    u'Dataset Agent CPU',
+    u'Container Agent CPU',
+    u'Containers Converged Within Limit',
+    u'Scenario Requests Within Limit',
+]
+
+
+def write_csv(results, filename):
+    with open(filename, 'w') as csvfile:
+        writer = csv.DictWriter(csvfile, fieldnames=HEADERS)
+        writer.writeheader()
+        writer.writerows(results)
+
+
+def mean(values):
+    if len(values) > 0:
+        return sum(values) / len(values)
+    return None
+
+
+def cputime_for_process(results, process):
+    """
+    Calculate the CPU time for a process running in a particular scenario.
+
+    By default this will return the `mean` result.
+    """
+    process_results = itertools.ifilter(
+        lambda r: r['metric']['type'] == 'cputime' and r['process'] == process,
+        results
+    )
+    values = [r['value'] / r['wallclock'] for r in process_results]
+    return mean(values)
+
+
+def wallclock_for_operation(results, operation):
+    """
+    Calculate the wallclock time for a process running in a particular
+    scenario.
+
+    By default this will return the `mean` result.
+    """
+    operation_results = itertools.ifilter(
+        lambda r: r['metric']['type'] == 'wallclock'
+        and r['operation']['type'] == operation,
+        results
+    )
+    values = [r['value'] for r in operation_results]
+    return mean(values)
+
+
+def container_convergence(results, seconds):
+    """
+    Calculate the percentage of containers that converge within a given
+    time period.
+    """
+    convergence_results = [
+        r for r in results if r['metric']['type'] == 'wallclock'
+        and r['operation']['type'] == 'create-container'
+    ]
+    num_convergences = len(convergence_results)
+    if num_convergences > 0:
+        convergences_within_limit = [
+            r for r in convergence_results if r['value'] < seconds
+        ]
+        return len(convergences_within_limit) / num_convergences
```

Comment:
This appears to divide an int by an int, which rounds down.  This would mean 19 / 20 returns 0 instead of 0.95. (Unless I'm missing something?)